File: scripts/extract-images.js
const XLSX = require('xlsx');
const fs = require('fs');
const path = require('path');

// Read the Excel file
const filePath = path.join(__dirname, '..', 'Final Vehicle Model Wise Specifications -01.xlsx');
const workbook = XLSX.readFile(filePath, { cellImages: true });

console.log('Extracting images from Excel file...\n');

// Create images directory if it doesn't exist
const imagesDir = path.join(__dirname, '..', 'public', 'vehicle-images');
if (!fs.existsSync(imagesDir)) {
  fs.mkdirSync(imagesDir, { recursive: true });
}

// Track extracted images
const extractedImages = [];

// Process each sheet
workbook.SheetNames.forEach((sheetName, sheetIndex) => {
  console.log(`Processing sheet ${sheetIndex + 1}/${workbook.SheetNames.length}: ${sheetName}`);
  
  const worksheet = workbook.Sheets[sheetName];
  
  // Extract make and model from sheet name
  const nameParts = sheetName.split(/[-_]/);
  const make = nameParts[0]?.trim() || 'Unknown';
  const model = nameParts.slice(1).join(' ').trim() || 'Unknown';
  
  // Normalize names for file naming
  const normalizeName = (name) => {
    return name
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, '-')
      .replace(/^-+|-+$/g, '');
  };
  
  const makeNormalized = normalizeName(make);
  const modelNormalized = normalizeName(model);
  
  // Check if worksheet has images
  if (worksheet['!images'] && worksheet['!images'].length > 0) {
    worksheet['!images'].forEach((image, imgIndex) => {
      try {
        // Generate filename
        const imageExt = image.extension || 'png';
        const filename = `${makeNormalized}-${modelNormalized}-${imgIndex + 1}.${imageExt}`;
        const filePath = path.join(imagesDir, filename);
        
        // Write image data
        fs.writeFileSync(filePath, image.data);
        
        extractedImages.push({
          sheet: sheetName,
          make,
          model,
          filename,
          path: `/vehicle-images/${filename}`,
          index: imgIndex
        });
        
        console.log(`  ✓ Extracted: ${filename}`);
      } catch (error) {
        console.error(`  ✗ Error extracting image ${imgIndex + 1} from ${sheetName}:`, error.message);
      }
    });
  } else {
    // Try alternative method - check for embedded objects
    if (worksheet['!drawings']) {
      console.log(`  ℹ Found drawings but no images array for ${sheetName}`);
    }
  }
});

console.log(`\n✅ Extracted ${extractedImages.length} images total`);
console.log(`📁 Images saved to: ${imagesDir}`);

// Save mapping for reference
const mappingPath = path.join(__dirname, '..', 'image-mapping.json');
fs.writeFileSync(mappingPath, JSON.stringify(extractedImages, null, 2));
console.log(`📝 Image mapping saved to: ${mappingPath}`);

File: scripts/extract_images.py
import openpyxl
import os
import json
from pathlib import Path

# Path to Excel file
excel_path = r"D:\specsheet-system\Final Vehicle Model Wise Specifications -01.xlsx"
output_dir = r"D:\specsheet-system\public\vehicle-images"

# Create output directory if it doesn't exist
os.makedirs(output_dir, exist_ok=True)

# Load workbook
print("Loading Excel file...")
wb = openpyxl.load_workbook(excel_path, data_only=True)

# Track extracted images
extracted_images = []

# Normalize name for file naming
def normalize_name(name):
    return name.lower().replace(' ', '-').replace('/', '-').replace('\\', '-').replace(':', '-').replace('*', '').replace('?', '').replace('"', '').replace('<', '').replace('>', '').replace('|', '').strip('-')

# Process each sheet
for sheet_name in wb.sheetnames:
    print(f"\nProcessing sheet: {sheet_name}")
    
    ws = wb[sheet_name]
    
    # Extract make and model from sheet name
    name_parts = sheet_name.replace('_', '-').split('-')
    make = name_parts[0].strip() if name_parts else 'Unknown'
    model = '-'.join(name_parts[1:]).strip() if len(name_parts) > 1 else 'Unknown'
    
    make_normalized = normalize_name(make)
    model_normalized = normalize_name(model)
    
    # Find images in the sheet
    if hasattr(ws, '_images') and ws._images:
        print(f"  Found {len(ws._images)} image(s)")
        
        for idx, img in enumerate(ws._images):
            try:
                # Get image data
                image_data = img._data()
                
                # Determine file extension
                ext = 'png'  # Default
                if hasattr(img, 'format') and img.format:
                    ext = img.format.lower()
                elif image_data[:4] == b'\x89PNG':
                    ext = 'png'
                elif image_data[:2] == b'\xff\xd8':
                    ext = 'jpg'
                elif image_data[:4] == b'GIF8':
                    ext = 'gif'
                
                # Generate filename
                filename = f"{make_normalized}-{model_normalized}-{idx + 1}.{ext}"
                filepath = os.path.join(output_dir, filename)
                
                # Save image
                with open(filepath, 'wb') as f:
                    f.write(image_data)
                
                extracted_images.append({
                    'sheet': sheet_name,
                    'make': make,
                    'model': model,
                    'filename': filename,
                    'path': f'/vehicle-images/{filename}',
                    'index': idx
                })
                
                print(f"  ✓ Extracted: {filename}")
                
            except Exception as e:
                print(f"  ✗ Error extracting image {idx + 1}: {str(e)}")
    else:
        # Try alternative method - check for embedded objects
        print(f"  No images found using _images attribute")
        
        # Check for drawings/shapes
        if hasattr(ws, 'drawings') and ws.drawings:
            print(f"  Found drawings, but image extraction may require different method")

print(f"\n✅ Extracted {len(extracted_images)} images total")
print(f"📁 Images saved to: {output_dir}")

# Save mapping
mapping_path = r"D:\specsheet-system\image-mapping.json"
with open(mapping_path, 'w', encoding='utf-8') as f:
    json.dump(extracted_images, f, indent=2, ensure_ascii=False)
print(f"📝 Image mapping saved to: {mapping_path}")

